File: Up/IoTMcu/HC595.cs
using System;
using System.Device.Gpio;
using System.Threading;

namespace IoTMcu
{
    class HC595
    {
        private const int MRES = 10;
        private const int MCLK = 10;
        private const int MLOCK = 10;
        private const int MOUT = 10;
        private const int M1RED = 10;
        private const int M2RED = 10;
        private const int M1BLU = 10;
        private const int M2BLU = 10;

        public HC595()
        {
            IoTMcuMain.GpioController.OpenPin(MRES, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(MCLK, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(MLOCK, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(MOUT, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(M1RED, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(M2RED, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(M1BLU, PinMode.Output);
            IoTMcuMain.GpioController.OpenPin(M2BLU, PinMode.Output);

            IoTMcuMain.GpioController.Write(MRES, PinValue.Low);
            IoTMcuMain.GpioController.Write(MCLK, PinValue.Low);
            Thread.Sleep(10);
            IoTMcuMain.GpioController.Write(MRES, PinValue.High);
        }

        public static void Reset()
        {
            IoTMcuMain.GpioController.Write(MRES, PinValue.Low);
            Thread.Sleep(10);
            IoTMcuMain.GpioController.Write(MRES, PinValue.High);
        }

        public static void SetOut(bool val)
        {
            IoTMcuMain.GpioController.Write(MOUT, val ? PinValue.Low : PinValue.High);
        }
        public static void SetDate(PinValue[] data1, PinValue[] data2,
            PinValue[] data3, PinValue[] data4, int size, bool both = true)
        {
            int local;

            for (local = 0; local < size; local++)
            {
                IoTMcuMain.GpioController.Write(M1BLU, data1[local]);
                IoTMcuMain.GpioController.Write(M1RED, data3[local]);
                if (both)
                {
                    IoTMcuMain.GpioController.Write(M2BLU, data2[local]);
                    IoTMcuMain.GpioController.Write(M2RED, data4[local]);
                }

                IoTMcuMain.GpioController.Write(MCLK, PinValue.Low);
                Thread.Sleep(TimeSpan.FromMilliseconds(10));
                IoTMcuMain.GpioController.Write(MCLK, PinValue.High);
            }
        }
        public static void Unlock()
        {
            IoTMcuMain.GpioController.Write(MLOCK, PinValue.High);
            Thread.Sleep(TimeSpan.FromMilliseconds(10));
            IoTMcuMain.GpioController.Write(MLOCK, PinValue.Low);
        }
    }
}

File: Src/src/HC595.c
#include <stdio.h>
#include <HC595.h>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "driver/gpio.h"
#include "sdkconfig.h"

void HC595Init()
{
    gpio_pad_select_gpio(MRES);
    gpio_pad_select_gpio(MLOCK);
    gpio_pad_select_gpio(MCLK);
    gpio_pad_select_gpio(M1RED);
    gpio_pad_select_gpio(M1BLU);
    gpio_pad_select_gpio(M2RED);
    gpio_pad_select_gpio(M2BLU);

    gpio_set_direction(MRES, GPIO_MODE_OUTPUT);
    gpio_set_direction(MLOCK, GPIO_MODE_OUTPUT);
    gpio_set_direction(MCLK, GPIO_MODE_OUTPUT);
    gpio_set_direction(M1RED, GPIO_MODE_OUTPUT);
    gpio_set_direction(M1BLU, GPIO_MODE_OUTPUT);
    gpio_set_direction(M2RED, GPIO_MODE_OUTPUT);
    gpio_set_direction(M2BLU, GPIO_MODE_OUTPUT);

    gpio_set_level(MRES, 1);
    gpio_set_level(MLOCK, 0);
    gpio_set_level(MCLK, 0);
    gpio_set_level(M1RED, 1);
    gpio_set_level(M1BLU, 1);
    gpio_set_level(M2RED, 1);
    gpio_set_level(M2BLU, 1);

    Reset();
}

void Reset()
{
    gpio_set_level(MRES, 0);
    vTaskDelay(10 / portTICK_PERIOD_MS);
    gpio_set_level(MRES, 1);
}

void UnLock()
{
    gpio_set_level(MLOCK, 0);
    ets_delay_us(10);
    gpio_set_level(MLOCK, 1);
}

void SetDataA(uint8_t *data1, uint8_t *data2, uint8_t size)
{
    uint8_t i;
    uint8_t local;
    uint8_t data_1;
    uint8_t data_2;

    for (local = 0; local < size; local++)
    {
        data_1 = data1[local];
        data_2 = data2[local];
        for (i = 0; i < 8; i++)
        {
            gpio_set_level(MCLK, 0);
            // gpio_set_level(M1BLU, 0);
            // gpio_set_level(M1RED, 1);
            gpio_set_level(M1RED, data_1 & 0x80 ? 1 : 0);
            gpio_set_level(M1BLU, data_2 & 0x80 ? 1 : 0);

            data_1 <<= 1;
            data_2 <<= 1;
            gpio_set_level(MCLK, 1);
        }
    }
}

void SetDataB(uint8_t *data1, uint8_t *data2, uint8_t *data3, uint8_t *data4, uint8_t size)
{
    uint8_t i;
    uint8_t local;
    uint8_t data_1;
    uint8_t data_2;
    uint8_t data_3;
    uint8_t data_4;
    for (local = 0; local < size; local++)
    {
        data_1 = data1[local];
        data_2 = data2[local];
        data_3 = data3[local];
        data_4 = data4[local];
        for (i = 0; i < 8; i++)
        {
            gpio_set_level(MCLK, 0);
            if (data_1 & 0x80)
            {
                gpio_set_level(M1RED, 1);
            }
            else
            {
                gpio_set_level(M1RED, 0);
            }

            if (data_2 & 0x80)
            {
                gpio_set_level(M1BLU, 1);
            }
            else
            {
                gpio_set_level(M1BLU, 0);
            }

            if (data_3 & 0x80)
            {
                gpio_set_level(M2RED, 1);
            }
            else
            {
                gpio_set_level(M2RED, 0);
            }

            if (data_4 & 0x80)
            {
                gpio_set_level(M2BLU, 1);
            }
            else
            {
                gpio_set_level(M2BLU, 0);
            }

            data_1 <<= 1;
            data_2 <<= 1;
            data_3 <<= 1;
            data_4 <<= 1;

            gpio_set_level(MCLK, 1);
        }
    }
}